pkg/apis/vpc/v1alpha1/subnet_types.go

```diff
@@ -35,17 +39,26 @@ type SubnetSpec struct {
 	// +kubebuilder:validation:MaxItems=2
 	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
 	IPAddresses []string `json:"ipAddresses,omitempty"`
-	// DHCPConfig DHCP configuration.
-	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
-	DHCPConfig DHCPConfig `json:"DHCPConfig,omitempty"`
+
+	// DHCP mode of a SubnetSet cannot switch from DHCPDeactivated to DHCPServer or DHCPRelay.
+	// If subnetDHCPConfig is not set, the DHCP mode is DHCPDeactivated by default.
+	// In order to enforce this rule, two XValidation rules are defined.
+	// The rule in SubnetSetSpec prevents the condition that subnetDHCPConfig is not set in
+	// old SubnetSetSpec while the new SubnetSetSpec specifies a field other than DHCPDeactivated.
+	// The rule in SubnetDHCPConfig prevents the mode changing from empty or
+	// DHCPDeactivated to DHCPServer or DHCPRelay.
+
+	// DHCP configuration for Subnet.
+	SubnetDHCPConfig SubnetDHCPConfig `json:"subnetDHCPConfig,omitempty"`
 }
 
 // SubnetStatus defines the observed state of Subnet.
 type SubnetStatus struct {
-	NetworkAddresses    []string    `json:"networkAddresses,omitempty"`
-	GatewayAddresses    []string    `json:"gatewayAddresses,omitempty"`
-	DHCPServerAddresses []string    `json:"DHCPServerAddresses,omitempty"`
-	Conditions          []Condition `json:"conditions,omitempty"`
+	NetworkAddresses    []string       `json:"networkAddresses,omitempty"`
+	GatewayAddresses    []string       `json:"gatewayAddresses,omitempty"`
+	DHCPServerAddresses []string       `json:"DHCPServerAddresses,omitempty"`
+	DHCPMode            DHCPConfigMode `json:"DHCPMode,omitempty"`
```

Comment:
do we still need DHCPMode here in SubnetStatus?
do we need dhcp_ranges showed?

Reply:
As discussed in chat, we will remove the DHCPMode and not add dhcp_ranges for now, thanks